### PR TITLE
Federated Reconstruction using Multimodal Data

### PR DESCRIPTION
**Pull request by minseok-ryu**
Federated Reconstruction using Multimodal Data

Here, we add an example and implement an application-driven algorithm that reconstructs an object from multimodal data created by Advanced Photon Source (APS) at Argonne.

- Each client has matrix "A" and vector "b" and aims to find "w" that minimizes ||Aw-b||^2.
- Linkage between clients exists; one client's "w" is a convex combination of the others, where all "w"s must be non-negative.
- It is a convex minimization problem with the convex combination constraint and the non-negativity restriction, which can be solved by the projected gradient descent algorithm. In the federated setting, it requires to solve a projection model at the server.
- In the proposed and implemented algorithm, the projection model is replaced with a set of simple mathematical operation.

### CODE REVIEW

**Comment by minseok-ryu** [comment]
To do list
- Compare the results with the ones produced by my local implementation
- Log 
- Tensorboard

File: src/appfl/config/fed/fedres.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+            ## tensorboard file name
+            "t_filename": "",
+            ## logging information
+            "logginginfo": {},
+            ## coefficient assigned for each client
+            "coeff": {},
+            ## ground truth (to compute the mean squared error for every iterations)
+            "w_truth": {0: [], 1: [], 2: [], 3: []},
+            ## Clients optimizer
+            "optim": "SGD",
```
Comment (kibaekkim) [inline src/appfl/config/fed/fedres.py:28]: Does it run with other optimizers?

**Review by kibaekkim — CHANGES_REQUESTED** [review]
This needs to be added to the CI for testing.

File: src/appfl/config/fed/fedres.py (hunk 1 above)
Comment (minseok-ryu) [reply src/appfl/config/fed/fedres.py:28]: Yes.

**Comment by minseok-ryu** [comment]
@kibaekkim For testing, do I need to upload data in github?

**Comment by kibaekkim** [comment]
> @kibaekkim For testing, do I need to upload data in github?

A toy-size test data should be sufficient.

File: examples/fedrecon_no_mpi.py [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+    """ User-defined data """
+    train_datasets = get_data(cfg)
+
+    """ User-defined model """
+    num_feature = train_datasets[0][0][0].size(dim=0)
+    model = LinearRegression(num_feature, 1)
+    loss_fn = torch.nn.MSELoss(reduction="sum")
+
+    """ Running """
+    rs.run_serial(cfg, model, loss_fn, train_datasets, Dataset(), args.dataset)
```
Comment (Zilinghan) [inline examples/fedrecon_no_mpi.py:149]: I wonder whether there is any specific reason only using serial run, or just for easy testing

Comment (minseok-ryu) [reply examples/fedrecon_no_mpi.py:149]: Just for easy testing. In this specific application, the number of clients is fixed to 4, so I think a serial run is good enough for now.

File: tests/test_aps.py
Comment (kibaekkim) [inline tests/test_aps.py:1]: To be consistent in naming, please change the file name (and the function names inside) to `test_fedres`.

File: setup.py [hunk 3]
```diff
@@ -57,6 +57,7 @@
             "pytest-mpi",
             "torchvision",
             "mpi4py",
+            "scipy",
```
Comment (kibaekkim) [inline setup.py:60]: The `scipy` package is used in the example and should be moved to under `examples`.

**Review by kibaekkim — COMMENTED** [review]
### One last comment. 
I realized that the code is written for a very specific application. The comments even mention application-specific terms like XRT. At this point, it is also not clear if the code can be generalized to a class of problems. This implementation looks more like an advanced use case of APPFL. 

**Before merging this PR**, please add a clear warning/description about that at the beginning of `algorithm/fedres.py`. Otherwise, the PR looks good.
